crates/networking/discv5/src/subnet.rs

```diff
@@ -1,110 +1,241 @@
+    fn from(bits: BitVector<U4>) -> Self {
+        Self(bits)
     }
 }
 
-impl Encodable for Subnets {
+impl Encodable for SyncCommitteeSubnets {
     fn encode(&self, out: &mut dyn BufMut) {
-        let ssz_bytes = self.as_ssz_bytes();
-        let bytes = Bytes::from(ssz_bytes);
+        let bytes = Bytes::from(self.0.as_ssz_bytes());
         bytes.encode(out);
     }
 }
 
-impl Decodable for Subnets {
+impl Decodable for SyncCommitteeSubnets {
     fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
         let bytes = Bytes::decode(buf)?;
-        let subnets = Subnets::from_ssz_bytes(&bytes).map_err(|err| {
+        let subnets = BitVector::<U4>::from_bytes(bytes.to_vec().into()).map_err(|err| {
             alloy_rlp::Error::Custom(Box::leak(
-                format!("Failed to decode SSZ subnets: {err:?}").into_boxed_str(),
+                format!("Failed to decode SSZ SyncCommitteeSubnets: {err:?}").into_boxed_str(),
             ))
         })?;
-        Ok(subnets)
+        Ok(Self(subnets))
+    }
+}
+
+impl AttestationSubnets {
+    pub fn enable_subnet(&mut self, subnet: Subnet) -> anyhow::Result<()> {
+        match subnet {
+            Subnet::Attestation(id) => {
+                if id < 64 {
+                    self.set(id as usize, true)?;
+                }
+                Ok(())
+            }
+            _ => Err(anyhow!(
+                "Cannot enable non-attestation subnet in AttestationSubnets"
+            )),
+        }
+    }
+
+    pub fn disable_subnet(&mut self, subnet: Subnet) -> anyhow::Result<()> {
+        match subnet {
+            Subnet::Attestation(id) => {
+                if id < 64 {
+                    self.set(id as usize, false)?;
+                }
+                Ok(())
+            }
+            _ => Err(anyhow!(
+                "Cannot disable non-attestation subnet in AttestationSubnets"
+            )),
+        }
+    }
+}
+
+impl SyncCommitteeSubnets {
+    pub fn enable_subnet(&mut self, subnet: Subnet) -> anyhow::Result<()> {
+        match subnet {
+            Subnet::SyncCommittee(id) => {
+                if id < SYNC_COMMITTEE_SUBNET_COUNT as u8 {
+                    self.set(id as usize, true)?;
+                }
+                Ok(())
+            }
+            _ => Err(anyhow!(
+                "Cannot enable non-sync committee subnet in SyncCommitteeSubnets"
+            )),
+        }
+    }
+
+    pub fn disable_subnet(&mut self, subnet: Subnet) -> anyhow::Result<()> {
+        match subnet {
+            Subnet::SyncCommittee(id) => {
+                if id < SYNC_COMMITTEE_SUBNET_COUNT as u8 {
+                    self.set(id as usize, false)?;
+                }
+                Ok(())
+            }
+            _ => Err(anyhow!(
+                "Cannot disable non-sync committee subnet in SyncCommitteeSubnets"
+            )),
+        }
     }
 }
 
 pub fn subnet_predicate(subnets: Vec<Subnet>) -> impl Fn(&Enr) -> bool + Send + Sync {
```

Comment:
could this be split into 2 different functions 1 for attestation subnets, and 1 for sync committees. As they should  be used in different cases no? And the current implementation has side affects which impacts each other.